src/backends/torch/torchlib.cc

```diff
@@ -485,10 +485,12 @@ namespace dd
                 TMLModel>::snapshot(int64_t elapsed_it, TorchSolver &tsolver)
   {
     this->_logger->info("Saving checkpoint after {} iterations", elapsed_it);
+    tsolver.swap_swa_sgd();
     this->_module.save_checkpoint(this->_mlmodel, std::to_string(elapsed_it));
-    // Save optimizer
+    // Save optimizer, swapped version
     tsolver.save(this->_mlmodel._repo + "/solver-" + std::to_string(elapsed_it)
                  + ".pt");
+    tsolver.swap_swa_sgd();
```

Comment:
This looks a bit like a hack

Reply:
agreed, but I could find any other way to do , do you have any idea ? 
(swa simply makes 

> Can it work with ADAM ?

theroetically yes, problem is on implementation side, the average is more naturally implemented inside the solver itself so that it can be saved/snapshoted along with other solver params, so we need to rewrite adam (or simply inherit) in order to add an attribute  and override save/load. 
Instead of inheriting a new solver from adam, I preferred to add this in our ranger (which is basically adam + some options). So if you want to try swa on adam, you can select ranger and deactivate all option (lookaeahd, gradient centralization and so on) in order to get vanilla adam

Reply:
I would suggest moving "swap_swa_sgd" call directly in ranger.
If we need to save the solver both with average and with gradients, create a new method "save_predict_checkpoint"? If the solver is always saved with averaged weights, no need for a new method.

Reply:
i cleanud up things a bit, now there is a solver.eval() and solver.train() that mimics module.train() and module.eval() , makeing things a bit more readable